service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -433,6 +472,21 @@ protected final ResourceLimitChecks getResourceLimitChecks() {
         return result;
     }
 
+    private void registerForLinkCloseEvent() {
+        if (this.metrics instanceof NoopBasedMetrics) {
+            return;
+        }
+
+        vertx.eventBus().consumer(AbstractHonoClient.EVENT_BUS_ADDRESS_CLIENT_SENDER_CLOSE, message -> {
```

Comment:
I don't understand the need to use the event bus here? Wouldn't it be possible to directly call this method?

Reply:
Currently the handler for link close events is not exposed by the client factories. This needs to be changed if we want to protocol adapters to be notified.
By doing so the API of the client would become more complex, because they hide the creation of the client in "getOrCreate" methods (e.g. [DownstreamSenderFactory#getTelemetrySender(String)](https://github.com/eclipse/hono/blob/ae47a9d25a13cf24695a7dfc35421087c3fd8861/client/src/main/java/org/eclipse/hono/client/DownstreamSenderFactory.java#L50)). Adding the close handler as a parameter to those methods would make them more error-prone, in that you must not provide a new handler instance for subsequent calls. 
Using the event bus decouples the client and the protocol adapter and I would rather remove the handler from the closeLinks() method. What downsides do you see in this solution?

Reply:
Ok, I see the problem. I agree that having this in the "getOrCreate" method is a bad idea.

The event bus has a tendency to make things: a) more complex b) error prune due to not having interfaces c) performance impacts vs a direct invocation … But I think you are right and in this case it is a good compromise.